impl/core/src/main/java/io/serverlessworkflow/impl/WorkflowUtils.java

```diff
@@ -221,11 +221,42 @@ public static final String checkSecret(
         .orElseThrow(() -> new IllegalStateException("Secret " + secretName + " does not exist"));
   }
 
-  public static URI concatURI(URI uri, String pathToAppend) {
-    return uri.getPath().endsWith("/")
-        ? uri.resolve(pathToAppend)
-        : URI.create(
-            uri.toString() + (pathToAppend.startsWith("/") ? pathToAppend : "/" + pathToAppend));
+  public static URI concatURI(URI base, String pathToAppend) {
+    if (pathToAppend == null || pathToAppend.isEmpty()) {
+      return base;
+    }
+
+    final URI child = URI.create(pathToAppend);
+    if (child.isAbsolute()) {
+      return child;
+    }
+
+    String basePath = base.getPath();
+    if (basePath == null || basePath.isEmpty()) {
+      basePath = "/";
+    } else if (!basePath.endsWith("/")) {
+      basePath = basePath + "/";
+    }
+
+    String relPath = child.getPath();
+    if (relPath == null) {
+      relPath = "";
+    }
+    while (relPath.startsWith("/")) {
+      relPath = relPath.substring(1);
+    }
+
+    String finalPath = basePath + relPath;
+
+    String query = child.getQuery();
+    String fragment = child.getFragment();
+
+    try {
+      return new URI(base.getScheme(), base.getAuthority(), finalPath, query, fragment);
+    } catch (Exception e) {
```

Comment:
```suggestion
    } catch (URISyntaxException e) {
```
Im not a fan of exception wrapping, but I guess the intention here is to to show a clear message to the user, in any case, if we are going to keep this block we should catch the specific exception, not all of them